code/src/terrat_files/github/sql/select_dirspaces_owned_by_other_pull_requests.sql

```diff
@@ -50,103 +50,94 @@ all_necessary_dirspaces as (
-          or (gwm.run_type in ('autoplan', 'plan') and gtp.has_changes is not null and not gtp.has_changes)
+          and ands.path is not null
+          and ands.workspace is not null
+          and (gpr.merged_at is not null
+               and (unlocks.unlocked_at is null or unlocks.unlocked_at < gpr.merged_at)
+               and (gwm.id is null
+                    or (gwm.state = 'completed'
+                        and (unlocks.unlocked_at is null or unlocks.unlocked_at < gwm.created_at)
+                        and (gwm.run_type in ('autoapply', 'apply', 'autoplan', 'plan'))
+                        and ands.lock_policy in ('strict', 'merge'))))
+               or (gpr.merged_at is null
+                   and gwm.id is not null
+                   and (unlocks.unlocked_at is null or unlocks.unlocked_at < gwm.created_at)
+                   and gwm.run_type in ('autoapply', 'apply')
+                   and ands.lock_policy in ('strict', 'apply'))
 ),
-unapplied_dirspaces as (
+dangling_dirspaces as (
     select
-        ands.repository as repository,
-        ands.pull_number as pull_number,
-        ands.path as path,
-        ands.workspace as workspace
-    from all_necessary_dirspaces as ands
-    inner join merged_pull_requests as merged
-        on merged.repository = ands.repository and merged.pull_number = ands.pull_number
-    left join applies_for_merged_pull_requests as applies
-        on applies.repository = ands.repository and applies.pull_number = ands.pull_number
-           and applies.path = ands.path and applies.workspace = ands.workspace
-    where applies.path is null and ands.lock_policy in ('strict', 'merge')
-),
--- And now combine our two lists of dirspaces:
---
--- 1. The dirspaces that a merged PR needs to be applied but haven't.
---
--- 2. The dirspaces for any PRs that have not been merged but have applies.
-all_dangling_dirspaces as (
-    select distinct
-        gpr.repository as repository,
-        gpr.pull_number as pull_number,
-        coalesce(uds.path, dds.path) as path,
-        coalesce(uds.workspace, dds.workspace) as workspace
-    from github_pull_requests as gpr
-    left join unapplied_dirspaces as uds
-        on uds.repository = gpr.repository and uds.pull_number = gpr.pull_number
-    left join dangling_dirspaces as dds
-        on dds.repository = gpr.repository and dds.pull_number = gpr.pull_number
-    where uds.repository is not null or dds.repository is not null
+        *
+    from incomplete_dirspaces_for_pull_requests as idspr
+    where idspr.rn = 1
+          and path is not null
+          and workspace is not null
+          and ((not is_merged
+                and run_type in ('autoapply', 'apply'))
+               or (is_merged
+                   and ((run_type in ('autoapply', 'apply') and not success)
+                        or (run_type in ('autoplan', 'plan') and success and has_changes))))
```

Comment:
This one is hard to parse, I wonder if there is a better way to write it,

```suggestion
          and ((not is_merged and run_type in ('autoapply', 'apply'))
               or (is_merged and not success and (run_type in ('autoapply', 'apply')))
               or (is_merged and success and has_changes and (run_type in ('autoapply', 'apply')))
```

although the suggestion is equally bad